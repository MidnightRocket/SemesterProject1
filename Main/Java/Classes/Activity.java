package worldOfZuul.Main.Java.Classes;

public class Activity {
    private String name;
    private int successPoints; // Positive number for number of points to receive when completing the activity.
    private int failurePoints; // Positive number to subtract from points when not completing or failing the activity.
    private int powerCost; // Amount of power required to complete the activity.
    private final boolean daily; // Determines if the task is daily (Daily=true, non-daily=false)
    private boolean activityDone; // False if activity is not done, true if activity is done.

    public Activity(String name, int successPoints, int failurePoints, int powerCost, boolean daily) {

        this.name = name;

        this.setSuccessPoints(successPoints);
        this.setFailurePoints(failurePoints);
        this.setPowerCost(powerCost);

        this.daily = daily;
    }

    public void setActivityDone() {
        activityDone = true;
    }

    public int getSuccessPoints() {
        return successPoints;
    }

    private void setSuccessPoints(int points) {
        if (points < 0) {
            throw new IllegalArgumentException("successPoints cannot be negative");
        } else {
            this.successPoints = points;
        }
    }

    public int getFailurePoints() {
        return failurePoints;
    }

    private void setFailurePoints(int points) {
        if (points < 0) {
            throw new IllegalArgumentException("failurePoints cannot be negative");
        } else {
            this.failurePoints = points;
        }
    }

    public int getPowerCost() {
        return powerCost;
    }

    private void setPowerCost(int powerCost) {
        if (powerCost < 0) {
            throw new IllegalArgumentException("powerCost cannot be negative");
        } else {
            this.powerCost = powerCost;
        }
    }

    public boolean isDaily() {
        return this.daily;
    }

    public boolean isActivityDone() {
        return this.activityDone;
    }

    public String getName() {
        return name;
    }
}