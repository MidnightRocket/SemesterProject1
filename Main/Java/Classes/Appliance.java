package worldOfZuul.Main.Java.Classes;
import worldOfZuul.Main.Java.Room;

public class Appliance {
    private final String name; // Name of the appliance

    private final Room roomReference; // Room the appliance is inside
    private final Activity activityReference; // Reference to associated activity

    public Appliance(String name, Room roomReference, Activity activityReference) {
        this.name = name;
        this.roomReference = roomReference;
        this.activityReference = activityReference;
    }
    public Room getRoomReference() {
        return roomReference;
    }
    public Activity getActivityReference() {
        return activityReference;
    }

    public String getName() {
        return name;
    }
}
