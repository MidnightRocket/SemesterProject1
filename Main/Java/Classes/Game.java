package worldOfZuul.Main.Java.Classes;

import java.util.Random;

public class Game {


    // Static attributes
    public static int day;
    public static int points;
    public static int power;

    // Instans attributes
    private int maxDays;
    private int defaultPower;
    public int extraPower;


    //Constructor
    public Game(int extraPower, int defaultPower, int maxDays) {
        this.day = 0;
        this.defaultPower = defaultPower;
        this.power = defaultPower;
        this.maxDays = maxDays;
        this.extraPower = extraPower;
    }


    //Metode for skift af dag
    public void nextDay() {
        day += 1;
        if (notLastDay()) {
            double randomExtraPower = Math.random() * extraPower;
            setPower(defaultPower + (int) randomExtraPower);
            extraPower -= (int) randomExtraPower;
        } else {
            setPower(defaultPower + extraPower);
            extraPower = 0;
        }
    }

    //Metode for tilføje til points værdi.
    public void addPoints(int pointsToAdd) {
        points += pointsToAdd;
    }

    //Metode for check day.
    public boolean notLastDay() {
        if (day < maxDays)
            return true;
        else return false;
    }

    //Metode getPower til at hente powerværdi.
    public int getPower() {
        return power;
    }

    //Metode for points værdi.
    public int getPoint() {
        return points;
    }

    //Metode set power værdi på dagen
    public void setPower(int newPower) {
        power = newPower;
    }
}