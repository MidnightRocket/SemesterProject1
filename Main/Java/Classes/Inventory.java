package worldOfZuul.Main.Java.Classes;

public class Inventory {
    private ArrayList<Item> listOfItems; // Dynamic array to hold items within an inventory list.

    public Inventory() {
        listOfItems = new ArrayList<Item>();
    }

    public ArrayList<Item> getInventoryList() {
        return listOfItems;
    }

    public void addItem(Item item) {
        listOfItems.add(item);
    }

    public void removeItem(Item item) {
        listOfItems.remove(item);
    }

    public void printItems() {
        if (ArrayList.isEmpty()) {
            System.out.println("Your inventory is currently empty!");
            
        }
        else {
            System.out.println("Your inventory currently contains: ");
            for (int i = 0; i < listOfItems.size(); i++) {
                System.out.println(listOfItems.get(i).getItemName());
            }
        }
    }
}
