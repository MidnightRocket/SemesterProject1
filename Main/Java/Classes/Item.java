package worldOfZuul.Main.Java.Classes;

/**
 * @param id                 Name of the item
 * @param applianceReference Reference to the associated appliance
 */
/*
 * Using record
 * https://dzone.com/articles/what-are-java-records
 * Records are class like constructs, but with immutable properties.
 * Methods can still be implemented if needed, as long they do not modifier the properties.
 * getter methods for each property is automatically generated.
 * *NOTE* the getter methods are named exactly the same as the property they refer to.
 */
public record Item(String id, String displayName) {
}