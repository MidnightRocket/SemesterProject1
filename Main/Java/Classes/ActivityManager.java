package worldOfZuul.Main.Java.Classes;

import worldOfZuul.Main.Java.Game;
import java.util.ArrayList;

public class ActivityManager {

    public ArrayList<Activity> listOfActivities;

    // Constructor WITH a list as input
    public ActivityManager(ArrayList<Activity> listOfActivities) {
        this.listOfActivities = listOfActivities;
    }

    // Constructor WITHOUT a list as input
    public ActivityManager() {
        this(new ArrayList<>());
    }


    public void addActivity(Activity activity) {
        listOfActivities.add(activity);
    }


    public void addPoints(Activity activity) {
        Game.points += activity.getSuccessPoints();
    }

    public void removePoints(Activity activity) {
        Game.points -= activity.getFailurePoints();
    }

    public void removePower(Activity activity) {
        Game.power -= activity.getPowerCost();
    }

    @Override
    public String toString() {
        String activityManagerString = "";
        if (listOfActivities.isEmpty()) {
            return "There are no more activitys.";
        } else {
            for (Activity activity : listOfActivities) {
                activityManagerString += " - " + activity.getName() + "\n";
            }
            return activityManagerString;
        }
    }
}
