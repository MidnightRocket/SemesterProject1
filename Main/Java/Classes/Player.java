package worldOfZuul.Main.Java.Classes;

import worldOfZuul.Main.Java.Command;
import worldOfZuul.Main.Java.Room;
import worldOfZuul.Main.Java.Game;

import java.util.ArrayList;

public class Player {
    Room currentRoom;
    Inventory inventoryReference;

    public Player(Inventory inventory) {
        this.currentRoom = this.getCurrentRoom();
        this.inventoryReference = inventory;
    }

    /*
     * If string matches an appliance in CurrentRoom, which has a complete-able activity "attached", and all criteria
     * for said activity (power amount, items in inv) is matched, the activity connected to the mentioned appliance will
     * be executed.
     */
    public void use(Command command){
        Room currentRoom = Game.getInstance().getCurrentRoom();
        String applianceName;
        Appliance applianceInUse;

        Item item = new Item("PlaceholderItem", new Appliance("PlaceholderAppliance",
                new Room("PlaceholderRoom"), new Activity("PlaceholderActivity",0,
                0,0,
                false))); // Placeholder item that will be replaced in inventory-search loop.

        Activity activity;

        applianceName = command.getCommandValue();
        // Checks whether the argument passed is a valid appliance name, and if that appliance has an active activity.
        if(currentRoom.getAppliance(applianceName).getActivityReference() == null){
            System.out.println("No appliance of that name has an activity to do in this room.");
            return;
        }


        applianceInUse = currentRoom.getAppliance(applianceName);
        activity = applianceInUse.getActivityReference();

        // Checks if the activity has already been done
        if(activity.isActivityDone()){
            if(activity.isDaily()) {
                System.out.println("This activity has already been performed today. Try again tomorrow.");
            }
            else {
                System.out.println("This activity has already been performed this week.");
            }
            return;
        }

        // Checks if the player has enough power
        if(Game.getInstance().getPower() < activity.getPowerCost()){
            String message = String.format("Insufficient power. This activity requires %d power.", activity.getPowerCost());
            System.out.println(message);
            return;
        }

        // Checks whether an Item inside inventory references the activity. (Might be changed later).
        boolean hasRequiredItem = false;
        for(int i = 0; i < getInventory().size(); i++){
            if(getInventory().get(i).getApplianceReference() == applianceInUse){
                item = getInventory().get(i);
                hasRequiredItem = true;
                break;
            }
        }
        if(!hasRequiredItem){
            System.out.println("An item is needed to perform this activity.");
            return;
        }

        // ACTIVITY IS COMPLETED SUCCESSFULLY
        Game.getInstance().addPoints(activity.getSuccessPoints()); // Add success points.
        Game.getInstance().removePower(activity.getPowerCost()); // Remove used power.
        activity.setActivityDone();
        getInventory().remove(item);

        String message = String.format("%d points has been added.", activity.getSuccessPoints());
        System.out.println(message);

        message = String.format("You have %d power left for the day.", Game.getInstance().getPower());
        System.out.println(message);
    }
    public void addItemToInventory(Item item) {
        inventoryReference.addItem(item);
    }

    public void removeItemFromInventory(Item item) {
        inventoryReference.removeItem(item);
    }

    public Room getCurrentRoom() {
        return Game.getInstance().getCurrentRoom();
    }

    public ArrayList<Item> getInventory() {
        return inventoryReference.getInventoryList();
    }
}
