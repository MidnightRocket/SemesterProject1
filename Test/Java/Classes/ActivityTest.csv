2,3,0,false
-2147483648,2147483647,-5,true
-500,1453,9,true
234,-23,-3,false
0,234,-3,true