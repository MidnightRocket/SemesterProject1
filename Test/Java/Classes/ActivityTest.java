package worldOfZuul.Test.Java.Classes;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import worldOfZuul.Main.Java.Classes.Activity;

import static org.junit.jupiter.api.Assertions.*;

class ActivityTest {

    @Test
    @DisplayName("Test Activity - Negative numbers in constructor 1")
    public void testConstructor1() {
        Activity activityTest = new Activity(-5, -15, -15, false);

        assertEquals(activityTest.getSuccessPoints(), 0);
        assertEquals(activityTest.getFailurePoints(), 0);
        assertEquals(activityTest.getPowerCost(), 0);

    }

    @Test
    @DisplayName("Test Activity - Negative numbers in constructor 2")
    public void testConstructor2() {
        Activity activityTest = new Activity(12345678, -378532, -12345678, false);

        assertEquals(activityTest.getSuccessPoints(), 12345678);
        assertEquals(activityTest.getFailurePoints(), 0);
        assertEquals(activityTest.getPowerCost(), 0);

    }

    @Test
    @DisplayName("Test Activity - Negative numbers in constructor 3")
    public void testConstructor3() {
        Activity activityTest = new Activity(-0, -2147483648, 2147483647, false);

        assertEquals(activityTest.getSuccessPoints(), 0);
        assertEquals(activityTest.getFailurePoints(), 0);
        assertEquals(activityTest.getPowerCost(), 2147483647);

    }

    @Test
    @DisplayName("Test Activity - Is daily getter")
    public void testIsDaily() {
        Activity activityTest1 = new Activity(-5, 5, -5, false);
        Activity activityTest2 = new Activity(-5, 5, -5, true);

        assertFalse(activityTest1.isDaily());
        assertTrue(activityTest2.isDaily());
    }
}