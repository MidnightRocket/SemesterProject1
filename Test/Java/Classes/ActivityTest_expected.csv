2,3,0,false
successPoints cannot be negative,2147483647,0,true
successPoints cannot be negative,1453,9,true
234,failurePoints cannot be negative,powerCost cannot be negative,false
0,234,powerCost cannot be negative,true